osidb/serializer.py

```diff
@@ -861,6 +862,69 @@ class Meta:
             + TrackingMixinSerializer.Meta.fields
         )
 
+    def update(self, affect, validated_data):
+        """
+        perform the tracker instance update
+        """
+
+        # NOTE: This overrides update() behavior provided by mixins, but mostly
+        #       mirrors the mixins. If mixins change, this method must be updated.
+        # NOTE: Look at TrackerSerializer.update() and
+        #       FlawPackageVersionSerializerMixin.update() for similar ways of
+        #       interposing actions within the mixins' update() MRO & flow.
+
+        ############################
+        # 1) prepare prerequisites #
+        ############################
+
+        # From ACLMixinSerializer.update()
+        validated_data = self.embargoed2acls(validated_data)
+
+        #########################
+        # 2) pre-update actions #
+        #########################
+
+        old_affectedness = affect.affectedness
+
+        #####################
+        # 3) update actions #
+        #####################
+
+        # From BugzillaSyncMixinSerializer.update()
+        for attr, value in validated_data.items():
+            setattr(affect, attr, value)
+
+        # From BugzillaBareSyncMixinSerializer.update()
+        affect.save(bz_api_key=self.get_bz_api_key())
+
+        ##########################
+        # 4) post-update actions #
+        ##########################
+
+        # When an affect goes from NEW affectedness to something else, any related tracker
+        # must have its validation-requested label removed, if it had one.
+        # Since the label is not stored in osidb, removal is done by regenerating the label
+        # by re-saving the tracker.
+
+        if (
+            old_affectedness != affect.affectedness
+            and old_affectedness == Affect.AffectAffectedness.NEW
+        ):
+            for tracker in affect.trackers.filter(type=Tracker.TrackerType.JIRA):
+                # do not raise validation errors here as the tracker is not what the user touches
+                # which would make the errors hard to understand
+                tracker.save(
+                    bz_api_key=self.get_bz_api_key(),
+                    jira_token=self.get_jira_token(),
+                    raise_validation_error=False,
+                )
+
+        #####################
+        # 5) return updated #
+        #####################
+
+        return affect
+
```

Comment:
I think this is a bad idea for two reasons:

1. The obvious one being that it's brittle, instead of depending on an abstraction's API (the mixins) we depend on its implementation details, I know we do this already elsewhere but that doesn't mean it should be the norm.
2. Business logic shouldn't really live this far up the stack (visual layer), I presume the reason for doing so is to have the user's JIRA API key, but do we really care about recording the person that changed the affectedness as the person setting/removing the label? I feel like an automated action like this one can be simply marked as "done by OSIDB" using OSIDB's JIRA API key, then this logic could live much more comfortably as a signal or within a save override, and in turn be less brittle.

WDYT?

Reply:
I agree that it does not seem right to have it there and it was me who started it. However I think that the audit is important so if we can do it on some other layer etc. then we can shift this logic there. Another reason is that on the serializer layer we have both the previous and the new object. If it is also simple to have it on the object level, then no problem to shift it there. Otherwise the comparison of what has changed would not be possible.

Reply:
I'll let you sort it out on the design level.

Low-level resolution suggestion: If Affect.save() always also requires jira key besides bz key, the custom logic could be moved there IMO. Would that be good enough for this PR?

Reply:
Let me be a bit more clear:

I am *against* having any api key logic further down the stack (even if it already is the case).
I am *for* having the label-setting logic further down the stack.

The latter implies that we don't track or audit _who_ triggered the setting of the label, because it is **unimportant**, setting the label is not something a user can do deliberately, the action of setting the label is not transparent to them, they are not aware of how or what triggers it, so why should they be liable for it? It's an automation from the system that may be triggered by a user's action, in this case changing an affect's affectedness for which we already have an audit trail.

TL;DR: unless we agree not to keep an audit of who set/unset the label, keep the current implementation.

Reply:
Can we then merge this PR? I think having the label-setting logic further down the stack would require reengineering that would be out of the scope of this task.